WordPress/src/main/java/org/wordpress/android/viewmodel/mlp/ModalLayoutPickerViewModel.kt

```diff
@@ -34,28 +46,62 @@ class ModalLayoutPickerViewModel @Inject constructor(
     private val _uiState: MutableLiveData<UiState> = MutableLiveData()
     val uiState: LiveData<UiState> = _uiState
 
+    private val contentUiState: ContentUiState
+        get() = uiState.value as? ContentUiState ?: ContentUiState()
+
     /**
      * Create new page event
      */
     private val _onCreateNewPageRequested = SingleLiveEvent<Unit>()
     val onCreateNewPageRequested: LiveData<Unit> = _onCreateNewPageRequested
 
+    init {
+        dispatcher.register(this)
+    }
+
+    override fun onCleared() {
+        dispatcher.unregister(this)
+        super.onCleared()
+    }
+
     fun init() {
-        updateUiState(ContentUiState())
-        loadLayouts()
-        loadCategories()
-        updateButtonsUiState()
+        fetchLayouts()
+    }
+
+    private fun fetchLayouts() {
+        updateUiState(LoadingUiState)
+        val siteId = appPrefsWrapper.getSelectedSite()
+        val site = siteStore.getSiteByLocalId(siteId)
+        if (site.isWPCom) {
+            dispatcher.dispatch(SiteActionBuilder.newFetchBlockLayoutsAction(site))
+        } else {
+            layouts = GutenbergPageLayoutFactory.makeDefaultPageLayouts()
+            loadLayouts()
+            loadCategories()
+        }
+    }
+
+    @Subscribe(threadMode = ThreadMode.MAIN)
+    fun onBlockLayoutsFetched(event: OnBlockLayoutsFetched) {
+        if (event.isError) {
+            updateUiState(ErrorUiState(event.error.message))
+        } else {
+            layouts = GutenbergPageLayouts(event.layouts, event.categories)
+            loadLayouts()
+            loadCategories()
+        }
     }
 
     private fun loadLayouts() {
-        val state = uiState.value as ContentUiState
+        val state = contentUiState
```

Comment:
We might want to consider moving this method to a bg/io thread as well, wdyt?

Reply:
As above fixed with 05ca9af80c966cdba96a30b4335a4a04bf31901e

Reply:
Same as above

Reply:
As above